rustls/src/client/hs.rs

```diff
@@ -38,6 +38,55 @@ use ring::digest::Digest;
 pub type NextState = Box<dyn State + Send + Sync>;
 pub type NextStateOrError = Result<NextState, TLSError>;
 
+/// The client-side handshake state machine
+///
+/// InitialState ->                             ExpectServerHello
+///                                             | ExpectServerHelloOrHelloRetryRequest
+///
+/// ExpectServerHello ->                        tls13::ExpectEncryptedExtensions
+///                                             | tls12::ExpectCertificate
+///                                             | tls12::ExpectNewTicket
+///                                             | tls12::ExpectCCS
+///
+/// ExpectServerHelloOrHelloRetryRequest ->     ExpectServerHello
+///
+/// tls12::ExpectCertificate ->                 tls12::ExpectCertificateStatusOrServerKX
+///                                             | tls12::ExpectServerKX
+///
+/// tls12::ExpectCertificateStatusOrServerKX -> tls12::ExpectServerKX
+///                                             | tls12::ExpectCertificateStatus
+///
+/// tls12::ExpectCertificateStatus ->           tls12::ExpectServerKX
+///
+/// tls12::ExpectServerKX ->                    tls12::ExpectServerDoneOrCertReq
+///
+/// tls12::ExpectServerDoneOrCertReq ->         tls12::ExpectCertificateRequest
+///                                             | tls12::ExpectServerDone
+///
+/// tls12::ExpectCertificateRequest ->          tls12::ExpectServerDone
+///
+/// tls12::ExpectServerDone ->                  tls12::ExpectNewTicket
+///                                             | tls12::ExpectCCS
+///
+/// tls12::ExpectNewTicket ->                   tls12::ExpectCCS
+///
+/// tls12::ExpectCCS ->                         tls12::ExpectFinished
+///
+/// tls12::ExpectFinished ->                    tls12::ExpectTraffic
+///
+/// tls13::ExpectEncryptedExtensions ->         tls13::ExpectFinished
+///                                             | tls13::ExpectCertificateOrCertReq
+///
+/// tls13::ExpectCertificateOrCertReq ->        tls13::ExpectCertificate
+///                                             | tls13::ExpectCertificateRequest
+///
+/// tls13::ExpectCertificateRequest ->          tls13::ExpectCertificate
+///
+/// tls13::ExpectCertificate ->                 tls13::ExpectCertificateVerify
+///
+/// tls13::ExpectCertificateVerify ->           tls13::ExpectFinished
+///
+/// tls13::ExpectFinished ->                    tls13::ExpectTraffic
```

Comment:
I don't think adding this comment is a good idea. This is just duplicating what's in the spec, it takes work to verify that the comment matches the spec, and it would take work to maintain the comment when the code changes (chances are very high the comment will get out of sync with the code). Instead we should just try to make the code match the diagrams in the spec more directly.